trace-forward/src/Trace/Forward/Network/Acceptor.hs

```diff
@@ -0,0 +1,193 @@
+import           Ouroboros.Network.Protocol.Handshake.Codec (cborTermVersionDataCodec,
+                                                             noTimeLimitsHandshake,
+                                                             timeLimitsHandshake)
+import           Ouroboros.Network.Protocol.Handshake.Unversioned (UnversionedProtocol (..),
+                                                                   UnversionedProtocolData (..),
+                                                                   unversionedHandshakeCodec,
+                                                                   unversionedProtocolDataCodec)
+import           Ouroboros.Network.Protocol.Handshake.Type (Handshake)
+import           Ouroboros.Network.Protocol.Handshake.Version (acceptableVersion,
+                                                               simpleSingletonVersions)
+import           Ouroboros.Network.Util.ShowProxy (ShowProxy(..))
+
+import qualified Trace.Forward.Protocol.Acceptor as Acceptor
+import qualified Trace.Forward.Protocol.Codec as Acceptor
+import           Trace.Forward.Protocol.Limits (byteLimitsTraceForward, timeLimitsTraceForward)
+import           Trace.Forward.Protocol.Type
+import           Trace.Forward.Queue (getTraceObjects)
+import           Trace.Forward.Configuration (AcceptorConfiguration (..), HowToConnect (..))
+
+listenToForwarder
+  :: (CBOR.Serialise lo,
+      ShowProxy lo,
+      Typeable lo)
+  =>  IOManager
+  -> AcceptorConfiguration lo
+  -> ([lo] -> IO ())
+  -> (NodeInfo -> IO ())
+  -> IO ()
+listenToForwarder iomgr config@AcceptorConfiguration{..} loHandler niHandler =
+  case forwarderEndpoint of
+    LocalPipe localPipe -> do
+      let snocket = localSnocket iomgr localPipe
+          address = localAddressFromPath localPipe
+      doListenToForwarder snocket address noTimeLimitsHandshake app
+    RemoteSocket host port -> do
+      listenAddress:_ <- Socket.getAddrInfo Nothing (Just $ T.unpack host) (Just $ show port)
+      let snocket = socketSnocket iomgr
+          address = Socket.addrAddress listenAddress
+      doListenToForwarder snocket address timeLimitsHandshake app
+ where
+  app =
+    OuroborosApplication $ \_connectionId _shouldStopSTM ->
+      [ MiniProtocol
+          { miniProtocolNum    = MiniProtocolNum 1
+          , miniProtocolLimits = MiniProtocolLimits { maximumIngressQueue = maxBound }
+          , miniProtocolRun    = acceptTraceObjects config loHandler niHandler
+          }
+      ]
+
+doListenToForwarder
+  :: Ord addr
+  => Snocket IO fd addr
+  -> addr
+  -> ProtocolTimeLimits (Handshake UnversionedProtocol Term)
+  -> OuroborosApplication 'ResponderMode addr LBS.ByteString IO Void ()
+  -> IO ()
+doListenToForwarder snocket address timeLimits app = do
+  networkState <- newNetworkMutableState
+  nsAsync <- async $ cleanNetworkMutableState networkState
+  clAsync <- async . void $
+    withServerNode
```

Comment:
The acceptor (run by an external process), was supposed to connect to the forwarder (run by the node), not the other way around, isn't it?

Reply:
By default - yes, the acceptor is connecting to the forwarder.

Reply:
By default - yes. But technically, `cardano-tracer` can work as `Initiator` (by default) as well as `Responder`: https://github.com/input-output-hk/cardano-node/blob/tracing-master/cardano-tracer/src/Cardano/Tracer/Acceptors.hs#L147-L159

Reply:
Do we need this flexibility? This could be a source of misconfiguration or integration errors.

Reply:
Yes, unfortunately we do @coot -- since we need the node to be the responder always.

So in the local case it is unavoidable.

Reply:
Yes, we need it. This idea was from Neil, and Serge can explain it better.

Reply:
So the node can run the acceptor as well? Is that the intention?